lib/client/api.go

```diff
@@ -3715,7 +3715,8 @@ func (tc *TeleportClient) applyProxySettings(proxySettings webclient.ProxySettin
 				proxySettings.DB.PostgresPublicAddr)
 		}
 		tc.PostgresProxyAddr = net.JoinHostPort(addr.Host(), strconv.Itoa(addr.Port(tc.WebProxyPort())))
-	case proxySettings.DB.PostgresListenAddr != "":
+		// Listen address port applies if set and not in TLS routing mode.
+	case proxySettings.DB.PostgresListenAddr != "" && !proxySettings.TLSRoutingEnabled:
```

Comment:
If Postgres listen address is specified then isn't the intent that Postgres listener is separate and client should connect to it even if TLS routing is enabled? @smallinsky @greedy52 Can you folks check as well?

Reply:
I don't think that this is intended behavior and by design in case of TLS Routing enabled the Web Port takes precedence over 
[tunnel_public_addr](https://github.com/gravitational/teleport/issues/8952), postgres_public_addr,  mysql_public_addrs or [kube_public_addr](https://github.com/gravitational/teleport/pull/8940#discussion_r748454319)

I don't think that we want to change this behavior because it will mean that if the  postgres_public_addr was provided and TLS Routing is enable the local client will try to connect with TLS Routing connection to raw postgres listener that cannot handle TLS Routing Layer. 

This logic when postgres_public_addrs is used instead of Web Port address make only sense in context of `tsh db config` command that is use in case of separate listener mode and in case of TLS Routing  the `tsh proxy db` command should be used. 

https://goteleport.com/docs/connect-your-client/gui-clients/#get-connection-information
![Screenshot 2023-03-09 at 11 00 26](https://user-images.githubusercontent.com/22402974/223987582-1fa72dfb-dcf9-48ed-aa37-7ba0ae8f2ada.png)

Reply:
Maybe I misunderstood the issue. `multiplex` is enabled, and Postgres listener address is set but NOT the Postgres public address. In my mind, using the Postgres listener port (e.g. `:3080`) is the correct behaviour for `tsh db config`, assuming it's reachable. If `teleport.example.com:3080` is not reachable, what the Postgres listener address is used for?

Reply:
I see. In TLS Routing mode logic we are using the web port where `tsh db config` is intended to be used only is separate listener mode but we don't forbid this call from multiplex mode so a client can still connect to the proxy using raw postgres connection without local proxy. In multiple mode, in ping response, we are setting all listeners to the webport address: https://github.com/gravitational/teleport/blob/806a568ada7c640a64eb60f718e7d56be34049ad/lib/service/proxy_settings.go#L100

So the response looks like: 
```json
  "proxy": {
    "kube": {
      "enabled": true,
      "listen_addr": "0.0.0.0:3080"
    },
    "ssh": {
      "listen_addr": "0.0.0.0:3080",
      "tunnel_listen_addr": "0.0.0.0:3080",
      "web_listen_addr": "0.0.0.0:3080",
      "public_addr": "my.ice-berg.dev:443"
    },
    "db": {
      "postgres_listen_addr": "0.0.0.0:3080",
      "mysql_listen_addr": "0.0.0.0:3080"
    },
    "tls_routing_enabled": true
  },
  ```
Where the postgres address is build from public_addr host:`my.ice-berg.dev"  and postgres_listen_addr port `3080` 

but in the case of Multiplex mode the WebPortHost should take precedence.

Reply:
Thanks for confirming!

Reply:
> I see. In TLS Routing mode logic we are using the web port where `tsh db config` is intended to be used only is separate listener mode but we don't forbid this call from multiplex mode so a client can still connect to the proxy using raw postgres connection without local proxy. In multiple mode, in ping response, we are setting all listeners to the webport address:
> 
> https://github.com/gravitational/teleport/blob/806a568ada7c640a64eb60f718e7d56be34049ad/lib/service/proxy_settings.go#L100
> 
> So the response looks like:
> 
> ```json
>   "proxy": {
>     "kube": {
>       "enabled": true,
>       "listen_addr": "0.0.0.0:3080"
>     },
>     "ssh": {
>       "listen_addr": "0.0.0.0:3080",
>       "tunnel_listen_addr": "0.0.0.0:3080",
>       "web_listen_addr": "0.0.0.0:3080",
>       "public_addr": "my.ice-berg.dev:443"
>     },
>     "db": {
>       "postgres_listen_addr": "0.0.0.0:3080",
>       "mysql_listen_addr": "0.0.0.0:3080"
>     },
>     "tls_routing_enabled": true
>   },
> ```
> 
> Where the postgres address is build from public_addr host:`my.ice-berg.dev` and postgres_listen_addr port `3080`
> 
> but in the case of Multiplex mode the WebPortHost should take precedence.

@smallinsky that makes sense (for this fix). If `mysql_listen_addr` and `postgres_listen_addr` are inheriting `multiplexAddr`, shouldn't `postgres_public_addr` and `postgres_public_addr` also inherit web public address? And this webport wouldn't work with MySQL through webport without local proxy?